enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java

```diff
@@ -791,10 +796,75 @@ private void markAsWaitingFor( ForsetiLockManager.Lock lock, ResourceType type,
             // after we've generated a description of it.
             if ( lock.detectDeadlock( id() ) != -1 )
             {
-                waitList.clear();
-                throw new DeadlockDetectedException( message );
+                // If the deadlock is real, then an owner of this lock must be (transitively) waiting on a lock that
+                // we own. So to verify the deadlock, we traverse the lock owners and their `waitingForLock` fields,
+                // to find a lock that has us among the owners.
+                // We only act upon the result of this method if the `tries` count is above some threshold. The reason
+                // is that the Lock.collectOwners, which is algorithm relies upon, is inherently racy, and so only
+                // reduces the probably of a false positive, but does not eliminate them.
+                if ( isDeadlockReal( lock, tries ) )
+                {
+                    // After checking several times, this really does look like a real deadlock.
+                    waitList.clear();
+                    waitingForLock = null;
+                    throw new DeadlockDetectedException( message );
+                }
             }
         }
+        waitingForLock = null;
+    }
+
+    private boolean isDeadlockReal( ForsetiLockManager.Lock lock, int tries )
+    {
+        Set<ForsetiLockManager.Lock> waitedUpon = new HashSet<>();
```

Comment:
It took me a long time to grok what this set was tracking; maybe something like `seenLocks` or `locksAlreadyExplored` or something that highlights that it's just making sure the algorithm isn't going back to re-examine locks it's already looked at (assuming I'm understanding this right?).

I can't see anything obviously wrong below, but it's also a bit tricky to follow.. maybe I'm becoming feeble minded from too much python programming ;)

FWIW, and independent of this PR: I think you very much were right that this algorithm overall optimizes for the wrong thing. The problem that we needed to solve in the first place was not making deadlock detection fast, it was making acquiring a lock that's free fast. 

I think the overhead of continuing trying to patch over the problems with Dreadlocks is much higher than the overhead of writing a new lock manager that maintains a regular RAG but has a good primitive for acquiring a lock that can be acquired without waiting.